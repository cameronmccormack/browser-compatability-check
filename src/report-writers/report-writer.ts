import fs from 'fs';
import path from 'path';
import jsonfile from 'jsonfile';
import pug from 'pug';
import { OverallReport } from '../types/compatibility';
import { OutputReportFile } from '../types/report-options';

const createDirectoryIfNotExists = (filepath: string): void => {
  const directoryPath = filepath.split('/').slice(0, -1).join('/');
  try {
    fs.mkdirSync(directoryPath, { recursive: true });
  } catch {
    // do nothing - it's fine if the directory already exists
  }
};

const copyStaticAssets = (fileLocation: string): void => {
  const staticFileDirectoryPath =
    fileLocation.split('/').slice(0, -1).join('/') + '/static';

  try {
    fs.mkdirSync(staticFileDirectoryPath);
  } catch {
    // do nothing - it's fine if the directory already exists
  }

  fs.copyFileSync(
    path.join(__dirname, '../../images/logo-text.svg'),
    `${staticFileDirectoryPath}/logo.svg`,
  );
};

export const writeCompatibilityReportFiles = (
  report: OverallReport,
  outputReportFileSpecifications: OutputReportFile[],
): void => {
  outputReportFileSpecifications.forEach((fileSpec) => {
    createDirectoryIfNotExists(fileSpec.location);
    switch (fileSpec.type) {
      case 'html':
        fs.writeFileSync(
          fileSpec.location,
          pug.compileFile(`${__dirname}/templates/report.pug`)(report),
        );
        copyStaticAssets(fileSpec.location);
        break;
      case 'json':
        jsonfile.writeFileSync(fileSpec.location, report, { spaces: 2 });
    }
  });
};
