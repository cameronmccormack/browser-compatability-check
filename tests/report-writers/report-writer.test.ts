import fs from 'fs';
import jsonfile from 'jsonfile';
import { writeCompatibilityReportFiles } from '../../src/report-writers/report-writer';
import {
  compatibleOverallReport,
  customRulesOverallReport,
  failedOverallReport,
  multipleReportsOverallReport,
  partiallyCompatibleOverallReport,
  summaryOnlyOverallReport,
  unknownFeatureOverallReport,
} from '../test-data/overall-reports';
import { OutputReportFile } from '../../src/types/report-options';
import {
  compatibleHtmlReportSnapshot,
  customRulesCompatibilityReportSnapshot,
  failedCompatibilityReportSnapshot,
  multipleReportsCompatibilityReportSnapshot,
  partiallyCompatibleReportSnapshot,
  summaryOnlyCompatibilityReportSnapshot,
  unknownFeatureCompatibilityReportSnapshot,
} from '../test-data/output-report-snapshots/html-reports';
import { OverallReport } from '../../src/types/compatibility';

afterEach(() => jest.restoreAllMocks());

describe('unit tests', () => {
  test('creates the expected number and type of reports', () => {
    const writeFileSyncSpy = jest.spyOn(fs, 'writeFileSync').mockReturnValue();
    const mkdirSyncSpy = jest.spyOn(fs, 'mkdirSync').mockReturnValue(undefined);
    const copyFileSyncSpy = jest.spyOn(fs, 'copyFileSync').mockReturnValue();
    const jsonfileWriteSpy = jest
      .spyOn(jsonfile, 'writeFileSync')
      .mockReturnValue();

    const fileSpecs: OutputReportFile[] = [
      { type: 'json', location: 'example/path1/file.json' },
      { type: 'json', location: 'example/path2/file.json' },
      { type: 'html', location: 'example/path3/file.html' },
      { type: 'json', location: 'example/path4/file.json' },
    ];

    writeCompatibilityReportFiles(compatibleOverallReport, fileSpecs);

    fileSpecs.forEach((spec) => {
      const outputDirectory = spec.location.split('/').slice(0, -1).join('/');

      if (spec.type === 'html') {
        expect(mkdirSyncSpy).toHaveBeenCalledWith(`${outputDirectory}/static`);
        expect(copyFileSyncSpy).toHaveBeenCalledWith(
          'images/logo-text.svg',
          `${outputDirectory}/static/logo.svg`,
        );
        expect(writeFileSyncSpy).toHaveBeenCalledWith(
          spec.location,
          expect.any(String),
        );
      }

      if (spec.type === 'json') {
        expect(jsonfileWriteSpy).toHaveBeenCalledWith(
          spec.location,
          compatibleOverallReport,
          { spaces: 2 },
        );
      }
    });
  });
});

type SnapshotTestData = {
  overallReport: OverallReport;
  expectedHtml: string;
};

const snapshotTestCases: [string, SnapshotTestData][] = [
  [
    'compatible report',
    {
      overallReport: compatibleOverallReport,
      expectedHtml: compatibleHtmlReportSnapshot,
    },
  ],
  [
    'partially compatible report',
    {
      overallReport: partiallyCompatibleOverallReport,
      expectedHtml: partiallyCompatibleReportSnapshot,
    },
  ],
  [
    'failed report',
    {
      overallReport: failedOverallReport,
      expectedHtml: failedCompatibilityReportSnapshot,
    },
  ],
  [
    'unknown feature report',
    {
      overallReport: unknownFeatureOverallReport,
      expectedHtml: unknownFeatureCompatibilityReportSnapshot,
    },
  ],
  [
    'custom rules report',
    {
      overallReport: customRulesOverallReport,
      expectedHtml: customRulesCompatibilityReportSnapshot,
    },
  ],
  [
    'summary only report',
    {
      overallReport: summaryOnlyOverallReport,
      expectedHtml: summaryOnlyCompatibilityReportSnapshot,
    },
  ],
  [
    'multiple CSS files report',
    {
      overallReport: multipleReportsOverallReport,
      expectedHtml: multipleReportsCompatibilityReportSnapshot,
    },
  ],
];

describe('HTML report snapshot tests', () => {
  test.each<[string, SnapshotTestData]>(snapshotTestCases)(
    'creates the expected HTML report for the case: %s',
    (_, { overallReport, expectedHtml }) => {
      let htmlReport: string = 'to be updated';

      jest.spyOn(fs, 'writeFileSync').mockImplementationOnce((_, data) => {
        htmlReport = data as string;
      });
      jest.spyOn(fs, 'mkdirSync').mockReturnValue(undefined);
      jest.spyOn(fs, 'copyFileSync').mockReturnValue();

      writeCompatibilityReportFiles(overallReport, [
        { location: 'example.html', type: 'html' },
      ]);

      expect(htmlReport).toEqual(expectedHtml);
    },
  );
});
