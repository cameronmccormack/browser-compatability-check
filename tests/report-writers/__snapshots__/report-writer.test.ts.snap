// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML report snapshot tests creates the expected HTML report for the case: compatible report 1`] = `
"<html lang="en-GB"><head><title>Kompat Report</title><link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css"/><style>body {
  font-family: Lato, sans-serif;
  background-color: #F6BF4F;
  color: #331F00;
}

li {
  width: fit-content;
}

dt {
  width: fit-content;
}

table {
  border-collapse: collapse;
  border: 4px solid #F6BF4F
}

th {
  text-align: left;
  background-color: #F6BF4F;
  padding: 0.25rem;
}

td {
  padding: 0.25rem;
}

.result {
  font-family: Lucida Console, monospace;
  font-weight: bold;
  padding: 0.25rem;
}

.bold {
  font-weight: bold;
}

.pass {
  color: white;
  background-color: #148A00;
}

.pass-cell {
  color: #148A00;
}

.warn {
  color: black;
  background-color: #FFB32B;
}

.warn-cell {
  color: black;
  background-color: #FFB32B;
}

.fail {
  color: white;
  background-color: #E51854;
}

.fail-cell {
  color: white;
  background-color: #E51854;
}

.heading-wrapper {
  margin: 1rem;
  text-align: center;
}

.summary-wrapper {
  margin: 1rem;
  padding: 0.5rem 2rem;
  background-color: #FFEECB;
  border-radius: 2rem;
}

.report-wrapper {
  margin: 2rem 0;
  padding: 0.5rem 2rem;
  background-color: white;
  border-radius: 2rem;
}

.table-wrapper {
  overflow-x: auto;
}

.code-wrapper {
  overflow-x: auto;
  margin: 1rem 0;
  width: fit-content;
  max-width: 100%;
}

.logo-heading {
  color: #331F00;
  letter-spacing: 0.5em;
  margin: 0.5rem;
  font-weight: 200;
}

.copyright {
  color: #331F00;
  margin: 0.5rem;
  font-size: 0.75em;
  letter-spacing: 0.25em;
}</style></head><body><main><div class="heading-wrapper"><img src="static/logo.svg" alt="Kompat Logo" width="288"/><h1 class="logo-heading">KOMPAT</h1><div class="copyright">COPYRIGHT &copy; 2023 CAMERON MCCORMACK</div></div><div class="summary-wrapper"><h2>Overall Summary: <span class="pass result">PASS</span></h2><div class="report-wrapper"><dl><dt class="pass result report-summary-status">PASS:</dt><dd class="bold"><a href="#report-0">example/filepath/eg-file.css</a></dd></dl></div></div><div class="summary-wrapper"><h2>Summary of All Stylesheets:</h2><div class="report-wrapper"><h3 id="report-0"><span class="pass result">PASS</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">4</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="pass-cell">compatible</td></td></tr><tr><td>selector:last-child<td class="pass-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="pass-cell">compatible</td></td></tr><tr><td>function:calc<td class="pass-cell">compatible</td></td></tr></tbody></table></div></div></div></main></body></html>"
`;

exports[`HTML report snapshot tests creates the expected HTML report for the case: custom rules report 1`] = `
"<html lang="en-GB"><head><title>Kompat Report</title><link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css"/><style>body {
  font-family: Lato, sans-serif;
  background-color: #F6BF4F;
  color: #331F00;
}

li {
  width: fit-content;
}

dt {
  width: fit-content;
}

table {
  border-collapse: collapse;
  border: 4px solid #F6BF4F
}

th {
  text-align: left;
  background-color: #F6BF4F;
  padding: 0.25rem;
}

td {
  padding: 0.25rem;
}

.result {
  font-family: Lucida Console, monospace;
  font-weight: bold;
  padding: 0.25rem;
}

.bold {
  font-weight: bold;
}

.pass {
  color: white;
  background-color: #148A00;
}

.pass-cell {
  color: #148A00;
}

.warn {
  color: black;
  background-color: #FFB32B;
}

.warn-cell {
  color: black;
  background-color: #FFB32B;
}

.fail {
  color: white;
  background-color: #E51854;
}

.fail-cell {
  color: white;
  background-color: #E51854;
}

.heading-wrapper {
  margin: 1rem;
  text-align: center;
}

.summary-wrapper {
  margin: 1rem;
  padding: 0.5rem 2rem;
  background-color: #FFEECB;
  border-radius: 2rem;
}

.report-wrapper {
  margin: 2rem 0;
  padding: 0.5rem 2rem;
  background-color: white;
  border-radius: 2rem;
}

.table-wrapper {
  overflow-x: auto;
}

.code-wrapper {
  overflow-x: auto;
  margin: 1rem 0;
  width: fit-content;
  max-width: 100%;
}

.logo-heading {
  color: #331F00;
  letter-spacing: 0.5em;
  margin: 0.5rem;
  font-weight: 200;
}

.copyright {
  color: #331F00;
  margin: 0.5rem;
  font-size: 0.75em;
  letter-spacing: 0.25em;
}</style></head><body><main><div class="heading-wrapper"><img src="static/logo.svg" alt="Kompat Logo" width="288"/><h1 class="logo-heading">KOMPAT</h1><div class="copyright">COPYRIGHT &copy; 2023 CAMERON MCCORMACK</div></div><div class="summary-wrapper"><h2>Overall Summary: <span class="fail result">FAIL</span></h2><div class="report-wrapper"><dl><dt class="pass result report-summary-status">PASS:</dt><dd class="bold"><a href="#report-0">example/filepath/eg-file.css</a></dd></dl></div></div><div class="summary-wrapper"><h2>Summary of All Stylesheets:</h2><div class="report-wrapper"><h3 id="report-0"><span class="pass result">PASS</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="fail-cell">4</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="fail-cell">compatible</td></td></tr><tr><td>selector:last-child<td class="fail-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="fail-cell">compatible</td></td></tr><tr><td>function:calc<td class="fail-cell">compatible</td></td></tr></tbody></table></div></div></div></main></body></html>"
`;

exports[`HTML report snapshot tests creates the expected HTML report for the case: failed report 1`] = `
"<html lang="en-GB"><head><title>Kompat Report</title><link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css"/><style>body {
  font-family: Lato, sans-serif;
  background-color: #F6BF4F;
  color: #331F00;
}

li {
  width: fit-content;
}

dt {
  width: fit-content;
}

table {
  border-collapse: collapse;
  border: 4px solid #F6BF4F
}

th {
  text-align: left;
  background-color: #F6BF4F;
  padding: 0.25rem;
}

td {
  padding: 0.25rem;
}

.result {
  font-family: Lucida Console, monospace;
  font-weight: bold;
  padding: 0.25rem;
}

.bold {
  font-weight: bold;
}

.pass {
  color: white;
  background-color: #148A00;
}

.pass-cell {
  color: #148A00;
}

.warn {
  color: black;
  background-color: #FFB32B;
}

.warn-cell {
  color: black;
  background-color: #FFB32B;
}

.fail {
  color: white;
  background-color: #E51854;
}

.fail-cell {
  color: white;
  background-color: #E51854;
}

.heading-wrapper {
  margin: 1rem;
  text-align: center;
}

.summary-wrapper {
  margin: 1rem;
  padding: 0.5rem 2rem;
  background-color: #FFEECB;
  border-radius: 2rem;
}

.report-wrapper {
  margin: 2rem 0;
  padding: 0.5rem 2rem;
  background-color: white;
  border-radius: 2rem;
}

.table-wrapper {
  overflow-x: auto;
}

.code-wrapper {
  overflow-x: auto;
  margin: 1rem 0;
  width: fit-content;
  max-width: 100%;
}

.logo-heading {
  color: #331F00;
  letter-spacing: 0.5em;
  margin: 0.5rem;
  font-weight: 200;
}

.copyright {
  color: #331F00;
  margin: 0.5rem;
  font-size: 0.75em;
  letter-spacing: 0.25em;
}</style></head><body><main><div class="heading-wrapper"><img src="static/logo.svg" alt="Kompat Logo" width="288"/><h1 class="logo-heading">KOMPAT</h1><div class="copyright">COPYRIGHT &copy; 2023 CAMERON MCCORMACK</div></div><div class="summary-wrapper"><h2>Overall Summary: <span class="warn result">WARN</span></h2><div class="report-wrapper"><dl><dt class="fail result report-summary-status">FAIL:</dt><dd class="bold"><a href="#report-0">example/filepath/eg-file.css</a></dd></dl></div></div><div class="summary-wrapper"><h2>Summary of All Stylesheets:</h2><div class="report-wrapper"><h3 id="report-0"><span class="fail result">FAIL</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">3</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="fail-cell">1</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="fail-cell">incompatible</td></td></tr><tr><td>selector:last-child<td class="pass-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="pass-cell">compatible</td></td></tr><tr><td>function:calc<td class="pass-cell">compatible</td></td></tr></tbody></table></div></div></div></main></body></html>"
`;

exports[`HTML report snapshot tests creates the expected HTML report for the case: multiple CSS files report 1`] = `
"<html lang="en-GB"><head><title>Kompat Report</title><link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css"/><style>body {
  font-family: Lato, sans-serif;
  background-color: #F6BF4F;
  color: #331F00;
}

li {
  width: fit-content;
}

dt {
  width: fit-content;
}

table {
  border-collapse: collapse;
  border: 4px solid #F6BF4F
}

th {
  text-align: left;
  background-color: #F6BF4F;
  padding: 0.25rem;
}

td {
  padding: 0.25rem;
}

.result {
  font-family: Lucida Console, monospace;
  font-weight: bold;
  padding: 0.25rem;
}

.bold {
  font-weight: bold;
}

.pass {
  color: white;
  background-color: #148A00;
}

.pass-cell {
  color: #148A00;
}

.warn {
  color: black;
  background-color: #FFB32B;
}

.warn-cell {
  color: black;
  background-color: #FFB32B;
}

.fail {
  color: white;
  background-color: #E51854;
}

.fail-cell {
  color: white;
  background-color: #E51854;
}

.heading-wrapper {
  margin: 1rem;
  text-align: center;
}

.summary-wrapper {
  margin: 1rem;
  padding: 0.5rem 2rem;
  background-color: #FFEECB;
  border-radius: 2rem;
}

.report-wrapper {
  margin: 2rem 0;
  padding: 0.5rem 2rem;
  background-color: white;
  border-radius: 2rem;
}

.table-wrapper {
  overflow-x: auto;
}

.code-wrapper {
  overflow-x: auto;
  margin: 1rem 0;
  width: fit-content;
  max-width: 100%;
}

.logo-heading {
  color: #331F00;
  letter-spacing: 0.5em;
  margin: 0.5rem;
  font-weight: 200;
}

.copyright {
  color: #331F00;
  margin: 0.5rem;
  font-size: 0.75em;
  letter-spacing: 0.25em;
}</style></head><body><main><div class="heading-wrapper"><img src="static/logo.svg" alt="Kompat Logo" width="288"/><h1 class="logo-heading">KOMPAT</h1><div class="copyright">COPYRIGHT &copy; 2023 CAMERON MCCORMACK</div></div><div class="summary-wrapper"><h2>Overall Summary: <span class="pass result">PASS</span></h2><div class="report-wrapper"><dl><dt class="pass result report-summary-status">PASS:</dt><dd class="bold"><a href="#report-0">example/filepath/eg-file.css</a></dd><dt class="pass result report-summary-status">PASS:</dt><dd class="bold"><a href="#report-1">example/filepath/eg-file.css</a></dd><dt class="pass result report-summary-status">PASS:</dt><dd class="bold"><a href="#report-2">example/filepath/eg-file.css</a></dd></dl></div></div><div class="summary-wrapper"><h2>Summary of All Stylesheets:</h2><div class="report-wrapper"><h3 id="report-0"><span class="pass result">PASS</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">4</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="pass-cell">compatible</td></td></tr><tr><td>selector:last-child<td class="pass-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="pass-cell">compatible</td></td></tr><tr><td>function:calc<td class="pass-cell">compatible</td></td></tr></tbody></table></div></div><div class="report-wrapper"><h3 id="report-1"><span class="pass result">PASS</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">4</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="pass-cell">compatible</td></td></tr><tr><td>selector:last-child<td class="pass-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="pass-cell">compatible</td></td></tr><tr><td>function:calc<td class="pass-cell">compatible</td></td></tr></tbody></table></div></div><div class="report-wrapper"><h3 id="report-2"><span class="pass result">PASS</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">4</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="pass-cell">compatible</td></td></tr><tr><td>selector:last-child<td class="pass-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="pass-cell">compatible</td></td></tr><tr><td>function:calc<td class="pass-cell">compatible</td></td></tr></tbody></table></div></div></div></main></body></html>"
`;

exports[`HTML report snapshot tests creates the expected HTML report for the case: partially compatible report 1`] = `
"<html lang="en-GB"><head><title>Kompat Report</title><link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css"/><style>body {
  font-family: Lato, sans-serif;
  background-color: #F6BF4F;
  color: #331F00;
}

li {
  width: fit-content;
}

dt {
  width: fit-content;
}

table {
  border-collapse: collapse;
  border: 4px solid #F6BF4F
}

th {
  text-align: left;
  background-color: #F6BF4F;
  padding: 0.25rem;
}

td {
  padding: 0.25rem;
}

.result {
  font-family: Lucida Console, monospace;
  font-weight: bold;
  padding: 0.25rem;
}

.bold {
  font-weight: bold;
}

.pass {
  color: white;
  background-color: #148A00;
}

.pass-cell {
  color: #148A00;
}

.warn {
  color: black;
  background-color: #FFB32B;
}

.warn-cell {
  color: black;
  background-color: #FFB32B;
}

.fail {
  color: white;
  background-color: #E51854;
}

.fail-cell {
  color: white;
  background-color: #E51854;
}

.heading-wrapper {
  margin: 1rem;
  text-align: center;
}

.summary-wrapper {
  margin: 1rem;
  padding: 0.5rem 2rem;
  background-color: #FFEECB;
  border-radius: 2rem;
}

.report-wrapper {
  margin: 2rem 0;
  padding: 0.5rem 2rem;
  background-color: white;
  border-radius: 2rem;
}

.table-wrapper {
  overflow-x: auto;
}

.code-wrapper {
  overflow-x: auto;
  margin: 1rem 0;
  width: fit-content;
  max-width: 100%;
}

.logo-heading {
  color: #331F00;
  letter-spacing: 0.5em;
  margin: 0.5rem;
  font-weight: 200;
}

.copyright {
  color: #331F00;
  margin: 0.5rem;
  font-size: 0.75em;
  letter-spacing: 0.25em;
}</style></head><body><main><div class="heading-wrapper"><img src="static/logo.svg" alt="Kompat Logo" width="288"/><h1 class="logo-heading">KOMPAT</h1><div class="copyright">COPYRIGHT &copy; 2023 CAMERON MCCORMACK</div></div><div class="summary-wrapper"><h2>Overall Summary: <span class="warn result">WARN</span></h2><div class="report-wrapper"><dl><dt class="warn result report-summary-status">WARN:</dt><dd class="bold"><a href="#report-0">example/filepath/eg-file.css</a></dd></dl></div></div><div class="summary-wrapper"><h2>Summary of All Stylesheets:</h2><div class="report-wrapper"><h3 id="report-0"><span class="warn result">WARN</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">3</td><td class="warn-cell">1</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="warn-cell">partial-support</td></td></tr><tr><td>selector:last-child<td class="pass-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="pass-cell">compatible</td></td></tr><tr><td>function:calc<td class="pass-cell">compatible</td></td></tr></tbody></table></div></div></div></main></body></html>"
`;

exports[`HTML report snapshot tests creates the expected HTML report for the case: summary only report 1`] = `
"<html lang="en-GB"><head><title>Kompat Report</title><link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css"/><style>body {
  font-family: Lato, sans-serif;
  background-color: #F6BF4F;
  color: #331F00;
}

li {
  width: fit-content;
}

dt {
  width: fit-content;
}

table {
  border-collapse: collapse;
  border: 4px solid #F6BF4F
}

th {
  text-align: left;
  background-color: #F6BF4F;
  padding: 0.25rem;
}

td {
  padding: 0.25rem;
}

.result {
  font-family: Lucida Console, monospace;
  font-weight: bold;
  padding: 0.25rem;
}

.bold {
  font-weight: bold;
}

.pass {
  color: white;
  background-color: #148A00;
}

.pass-cell {
  color: #148A00;
}

.warn {
  color: black;
  background-color: #FFB32B;
}

.warn-cell {
  color: black;
  background-color: #FFB32B;
}

.fail {
  color: white;
  background-color: #E51854;
}

.fail-cell {
  color: white;
  background-color: #E51854;
}

.heading-wrapper {
  margin: 1rem;
  text-align: center;
}

.summary-wrapper {
  margin: 1rem;
  padding: 0.5rem 2rem;
  background-color: #FFEECB;
  border-radius: 2rem;
}

.report-wrapper {
  margin: 2rem 0;
  padding: 0.5rem 2rem;
  background-color: white;
  border-radius: 2rem;
}

.table-wrapper {
  overflow-x: auto;
}

.code-wrapper {
  overflow-x: auto;
  margin: 1rem 0;
  width: fit-content;
  max-width: 100%;
}

.logo-heading {
  color: #331F00;
  letter-spacing: 0.5em;
  margin: 0.5rem;
  font-weight: 200;
}

.copyright {
  color: #331F00;
  margin: 0.5rem;
  font-size: 0.75em;
  letter-spacing: 0.25em;
}</style></head><body><main><div class="heading-wrapper"><img src="static/logo.svg" alt="Kompat Logo" width="288"/><h1 class="logo-heading">KOMPAT</h1><div class="copyright">COPYRIGHT &copy; 2023 CAMERON MCCORMACK</div></div><div class="summary-wrapper"><h2>Overall Summary: <span class="pass result">PASS</span></h2><div class="report-wrapper"><dl><dt class="pass result report-summary-status">PASS:</dt><dd class="bold"><a href="#report-0">example/filepath/eg-file.css</a></dd></dl></div></div><div class="summary-wrapper"><h2>Summary of All Stylesheets:</h2><div class="report-wrapper"><h3 id="report-0"><span class="pass result">PASS</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">4</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="pass-cell">None</li></ul></div></div></main></body></html>"
`;

exports[`HTML report snapshot tests creates the expected HTML report for the case: unknown feature report 1`] = `
"<html lang="en-GB"><head><title>Kompat Report</title><link href="https://fonts.googleapis.com/css?family=Lato:400,700" rel="stylesheet" type="text/css"/><style>body {
  font-family: Lato, sans-serif;
  background-color: #F6BF4F;
  color: #331F00;
}

li {
  width: fit-content;
}

dt {
  width: fit-content;
}

table {
  border-collapse: collapse;
  border: 4px solid #F6BF4F
}

th {
  text-align: left;
  background-color: #F6BF4F;
  padding: 0.25rem;
}

td {
  padding: 0.25rem;
}

.result {
  font-family: Lucida Console, monospace;
  font-weight: bold;
  padding: 0.25rem;
}

.bold {
  font-weight: bold;
}

.pass {
  color: white;
  background-color: #148A00;
}

.pass-cell {
  color: #148A00;
}

.warn {
  color: black;
  background-color: #FFB32B;
}

.warn-cell {
  color: black;
  background-color: #FFB32B;
}

.fail {
  color: white;
  background-color: #E51854;
}

.fail-cell {
  color: white;
  background-color: #E51854;
}

.heading-wrapper {
  margin: 1rem;
  text-align: center;
}

.summary-wrapper {
  margin: 1rem;
  padding: 0.5rem 2rem;
  background-color: #FFEECB;
  border-radius: 2rem;
}

.report-wrapper {
  margin: 2rem 0;
  padding: 0.5rem 2rem;
  background-color: white;
  border-radius: 2rem;
}

.table-wrapper {
  overflow-x: auto;
}

.code-wrapper {
  overflow-x: auto;
  margin: 1rem 0;
  width: fit-content;
  max-width: 100%;
}

.logo-heading {
  color: #331F00;
  letter-spacing: 0.5em;
  margin: 0.5rem;
  font-weight: 200;
}

.copyright {
  color: #331F00;
  margin: 0.5rem;
  font-size: 0.75em;
  letter-spacing: 0.25em;
}</style></head><body><main><div class="heading-wrapper"><img src="static/logo.svg" alt="Kompat Logo" width="288"/><h1 class="logo-heading">KOMPAT</h1><div class="copyright">COPYRIGHT &copy; 2023 CAMERON MCCORMACK</div></div><div class="summary-wrapper"><h2>Overall Summary: <span class="warn result">WARN</span></h2><div class="report-wrapper"><dl><dt class="fail result report-summary-status">FAIL:</dt><dd class="bold"><a href="#report-0">example/filepath/eg-file.css</a></dd></dl></div></div><div class="summary-wrapper"><h2>Summary of All Stylesheets:</h2><div class="report-wrapper"><h3 id="report-0"><span class="fail result">FAIL</span> example/filepath/eg-file.css</h3><h4>High-level Summary</h4><div class="table-wrapper"><table><thead><tr><th>Browser</th><th>Compatible</th><th>Partial Support</th><th>Flagged</th><th>Incompatible</th><th>Unknown</th></tr></thead><tbody><tr><td>chrome</td><td class="pass-cell">4</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td><td class="pass-cell">0</td></tr></tbody></table></div><h4>Unknown Features:</h4><ul><li class="fail-cell">property:not-a-real-feature:20px</li></ul><h4>Per-feature Summary</h4><div class="table-wrapper"><table><thead><tr><th>Rule Identifier<th>chrome</th></th></tr></thead><tbody><tr><td>property:color:red<td class="pass-cell">compatible</td></td></tr><tr><td>selector:last-child<td class="pass-cell">compatible</td></td></tr><tr><td>at-rule:charset<td class="pass-cell">compatible</td></td></tr><tr><td>function:calc<td class="pass-cell">compatible</td></td></tr></tbody></table></div></div></div></main></body></html>"
`;
